File: templane-ts/src/ir-generator.ts
import { ASTNode, Condition, TIRNode, TIRResult } from './models';

function resolve(data: Record<string, unknown>, path: string): unknown {
  let current: unknown = data;
  for (const part of path.split('.')) {
    if (typeof current !== 'object' || current === null || Array.isArray(current)) return null;
    const obj = current as Record<string, unknown>;
    if (!(part in obj)) return null;
    current = obj[part];
  }
  return current;
}

function evaluate(condition: Condition, data: Record<string, unknown>): boolean {
  if (condition.op === '==') {
    return String(resolve(data, condition.left)) === String(condition.right);
  }
  return false;
}

export function generate(
  astNodes: ASTNode[],
  data: Record<string, unknown>,
  schemaId: string,
  templateId: string,
): TIRResult {
  return {
    template_id: templateId,
    schema_id: schemaId,
    nodes: astNodes.map(n => nodeToTir(n, data)),
  };
}

function nodeToTir(node: ASTNode, data: Record<string, unknown>): TIRNode {
  switch (node.kind) {
    case 'text':
      return { kind: 'text', content: node.content };
    case 'expr':
      return { kind: 'expr', field: node.field, resolved: resolve(data, node.field) };
    case 'if': {
      const cond = evaluate(node.condition, data);
      const branch = cond ? node.then_branch : node.else_branch;
      return { kind: 'if', condition: cond, branch: branch.map(n => nodeToTir(n, data)) };
    }
    case 'foreach': {
      const itemsVal = resolve(data, node.iterable);
      const items: unknown[] = Array.isArray(itemsVal) ? itemsVal : [];
      return {
        kind: 'foreach',
        var: node.var,
        items: items.map(item => {
          const scope = { ...data, [node.var]: item };
          return node.body.map(n => nodeToTir(n, scope));
        }),
      };
    }
  }
}

File: templane-go/core/ir_generator.go
package core

import (
	"fmt"
	"strings"
)

// Generate walks an AST with data, producing a Typed Intermediate Representation.
// Missing paths resolve to nil; only "==" conditions are supported.
func Generate(ast []ASTNode, data map[string]any, schemaID, templateID string) TIRResult {
	nodes := make([]TIRNode, 0, len(ast))
	for _, n := range ast {
		nodes = append(nodes, nodeToTIR(n, data))
	}
	return TIRResult{TemplateID: templateID, SchemaID: schemaID, Nodes: nodes}
}

func nodeToTIR(node ASTNode, data map[string]any) TIRNode {
	switch node.Kind {
	case "text":
		return TIRNode{Kind: "text", Content: node.Content}
	case "expr":
		return TIRNode{Kind: "expr", Field: node.Field, Resolved: resolve(data, node.Field)}
	case "if":
		var cond bool
		if node.Condition != nil {
			cond = evaluate(*node.Condition, data)
		}
		var branch []ASTNode
		if cond {
			branch = node.ThenBranch
		} else {
			branch = node.ElseBranch
		}
		out := make([]TIRNode, 0, len(branch))
		for _, b := range branch {
			out = append(out, nodeToTIR(b, data))
		}
		return TIRNode{Kind: "if", Condition: cond, Branch: out}
	case "foreach":
		itemsVal := resolve(data, node.Iterable)
		items, _ := itemsVal.([]any)
		rendered := make([][]TIRNode, 0, len(items))
		for _, item := range items {
			scope := make(map[string]any, len(data)+1)
			for k, v := range data {
				scope[k] = v
			}
			scope[node.Var] = item
			body := make([]TIRNode, 0, len(node.Body))
			for _, b := range node.Body {
				body = append(body, nodeToTIR(b, scope))
			}
			rendered = append(rendered, body)
		}
		return TIRNode{Kind: "foreach", Var: node.Var, Items: rendered}
	}
	return TIRNode{}
}

func resolve(data map[string]any, path string) any {
	var current any = data
	for _, part := range strings.Split(path, ".") {
		m, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		v, present := m[part]
		if !present {
			return nil
		}
		current = v
	}
	return current
}

func evaluate(c Condition, data map[string]any) bool {
	if c.Op == "==" {
		return fmt.Sprintf("%v", resolve(data, c.Left)) == fmt.Sprintf("%v", c.Right)
	}
	return false
}
